tools/pythonpkg/src/python_udf.cpp

```diff
@@ -0,0 +1,362 @@
+			}
+
+			// Call the function
+			PyObject *ret = nullptr;
+			ret = PyObject_CallObject(function, bundled_parameters.ptr());
+			if (ret == nullptr && PyErr_Occurred()) {
+				if (exception_handling == PythonExceptionHandling::FORWARD_ERROR) {
+					auto exception = py::error_already_set();
+					throw InvalidInputException("Python exception occurred while executing the UDF: %s",
+					                            exception.what());
+				} else if (exception_handling == PythonExceptionHandling::RETURN_NULL) {
+					PyErr_Clear();
+					ret = Py_None;
+				} else {
+					throw NotImplementedException("Exception handling type not implemented");
+				}
+			}
+			python_objects.push_back(py::handle(ret));
+			python_results.push_back(ret);
+		}
+
+		// Cast the resulting native python to DuckDB, using the return type
+		// result.Resize(input.size());
+		NumpyScan::ScanObjectColumn(python_results.data(), input.size(), 0, result);
+		if (input.AllConstant()) {
+			result.SetVectorType(VectorType::CONSTANT_VECTOR);
+		}
+	};
+	return func;
+}
+
+static ScalarFunction CreateUDFInternal(const string &name, scalar_function_t func, const py::object &udf,
+                                        const py::object &parameters, const shared_ptr<DuckDBPyType> &return_type,
+                                        bool varargs, FunctionNullHandling null_handling) {
+	PythonUDFData data(name, std::move(func), varargs, null_handling);
+
+	data.AnalyzeSignature(udf);
+	data.OverrideParameters(parameters);
+	data.OverrideReturnType(return_type);
+	data.Verify();
+
+	return data.GetFunction();
+}
+
+ScalarFunction DuckDBPyConnection::CreatePyArrowScalarUDF(const string &name, const py::object &udf,
+                                                          const py::object &parameters,
+                                                          const shared_ptr<DuckDBPyType> &return_type, bool varargs,
+                                                          FunctionNullHandling null_handling,
+                                                          PythonExceptionHandling exception_handling) {
+	scalar_function_t func = CreateVectorizedFunction(udf.ptr(), exception_handling);
+	return CreateUDFInternal(name, std::move(func), udf, parameters, return_type, varargs, null_handling);
+}
+
+ScalarFunction DuckDBPyConnection::CreateScalarUDF(const string &name, const py::object &udf,
+                                                   const py::object &parameters,
+                                                   const shared_ptr<DuckDBPyType> &return_type, bool varargs,
+                                                   FunctionNullHandling null_handling,
+                                                   PythonExceptionHandling exception_handling) {
+	scalar_function_t func = CreateNativeFunction(udf.ptr(), exception_handling);
+	return CreateUDFInternal(name, std::move(func), udf, parameters, return_type, varargs, null_handling);
+}
```

Comment:
It seems to me that a lot of the code from here to the top of the API is duplicated functions and signatures of `CreateScalarUDF` and `CreatePyArrowScalarUDF`.

Can't we analyze the function and, if it uses Pyarrow immediately created the vectorized udf?
(No clue if the following works, just as an example)
e.g.,
```cpp
bool IsPyArrowFunction(const py::object& obj) {
    if (py::isinstance<py::function>(obj)) {
        auto func = obj.cast<py::function>();
        auto cpp_func = func.cpp_function();
        auto signature = cpp_func.signature();
        auto pos = signature.find("pybind11::detail::arrow_array");
        return (pos != std::string::npos);
    }
    return false;
}
```

Reply:
I thought we could, by using the type annotation, but we already use the type annotation to infer the parameter types.
It's possible to add support for both, and require passing explicit parameters if we used the type annotation to infer that the function is vectorized, is that a good compromise?

Reply:
Also, is vectorized a good name? I'm assuming both are vectorized, no?
The difference is the input/output of the UDFs. The name should probably reflect that.

Reply:
Hmm I agree, maybe I could turn it into `type: PythonUDFType`, with `PythonUDFType` being an enum: NATIVE, ARROW.
The option defaults to NATIVE then